docs/RFCS/20190906_flow_control.md

```diff
@@ -0,0 +1,137 @@
+
+### Congestion control framework
+
+The congestion control framework is separated into three parts, the congestion detect component, the data flow control component and the concurrency control component.
+
+#### The congestion detect component
+
+The congestion detect component will collect the key indices in DM-worker as following:
+
+- data loss rate: we will add a sliding window to record the count of downstream resource busy error (such as `tmysql.ErrPDServerTimeout`, `tmysql.ErrTiKVServerTimeout`, `tmysql.ErrTiKVServerBusy` etc) and the count of the distributed transactions. The data loss rate will be calculated with these two counters.
+- tps detection: we will add a sliding window to record the count of successfully executed transactions, and calculate tps with this value.
+- latency: transaction execution latency is already recorded in load and sync unit, we can use it directly.
+
+#### The data flow control component
+
+This component is responsible to give a reasonable concurrency value based on the data flow stage, current and historical data flow indices. Firstly we define some stage, which is used to represent the concurrency change stage.
+
+```go
+// CongestionStage represents the stage in a congestion control algorithm
+type CongestionStage int
+
+// Congestion stage list
+const (
+    // When the task starts with a low concurrency, the concurrency can be increased fast in this stage.
+    StageSlowStart CongestionStage = iota + 1
+    // When the concurrency exceeds a threshold, we recommend to increase slowly, which is often called the avoidance stage.
+    StageAvoidance
+    // The steady stage means the data flow tps reaches the bandwidth, the latency keeps steady and no data loss.
+    StageSteady
+    // When congestion happens and the concurrency is turned down, we enter the recovery stage.
+    StageRecovery
+)
+```
+
+We define a `FlowControl` interface, which is used to give a reasonable concurrency value based on the congestion stage, the historical and current data flow indices. We can use different congestion algorithm, just to implement the `FlowControl` interface.
+
+```go
+type CongestionIndictor struct {
+	DataLossRate float64
+	Tps          int
+	Latency      float64
+}
+
+type FlowControl interface {
+	// NextConcurrency gives the recommended concurrency value
+	NextConcurrency(stage CongestionStage, history []*CongestionIndictor, current *CongestionIndictor) int
+}
+```
+
+#### The concurrency control component
+
+This component is used to control the concurrency of DM work unit. Currently the code in load unit and sync unit don't support changing `pool-size` or `worker-count` dynamically very well. We have two ways to solve the problem.
+
+1. re-design the concurrent work model and support dynamically changing real worker count.
+2. create enough workers when the load/sync unit initializes, and control the concurrency by controlling the working worker number.
```

Comment:
do not forget to ensure `queueBucket = int(utils.GenHashKey(job.key)) % s.cfg.WorkerCount` work corrent.

no need to address this.